### PR TITLE
Fix invalid shape issue in random.uniform

### LINKED ISSUE #34363: random.uniform((),minval,maxval) returns array instead of scalar tensor when min or maxval is not a scalar tensor
```
>>> import tensorflow as tf
>>> scalar = tf.zeros(shape=())
>>> array = tf.zeros(shape=(1,))

>>> tf.random.uniform(shape=(),minval = scalar)
<tf.Tensor: id=25, shape=(), dtype=float32, numpy=0.021499991>

>>> tf.random.uniform(shape=(),minval = array)
<tf.Tensor: id=31, shape=(1,), dtype=float32, numpy=array([0.9388697], dtype=float32)>
```
Expected behavior is to either trow an error or treat single element tensor as scalar and return a scalar.

-win10, tf2, cuda

### PR DESCRIPTION
**Pull request by yongtang**
Note: This PR is a resubmission from #34399

This PR tries to address the issue raised in #34363 where
invalid shape passed to minval/maxval (expected to be 0-D)
does not raise an error.

The issue was that in most of the scenarios the shape was
checked inside the C++ kernel ops.

However, in one condition math_ops.add was used which will
implicitly do broadcast when necessarily.
This results in maxval/minval's shape getting carried.

This PR adds the shape check before math_ops.add, to make
sure the shape is guaranteed.

This PR fixes #34363.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

File: tensorflow/python/ops/random_ops.py [hunk 1]
```diff
@@ -300,6 +300,8 @@ def random_uniform(shape,
         if not maxval_is_one:
           result = math_ops.multiply(result, maxval)
       else:
+        maxval = array_ops.ensure_shape(maxval, ())
```
Comment (alextp) [inline tensorflow/python/ops/random_ops.py:303]: Doesn't this prevent us from making broadcast-correct randomuniform? As in, if I'm asking for a random uniform with shape [a, b] I should be able to have maxval/minval with any shape which broadcasts to [a, b] as we can broadcast the elementwise operations

**Comment by yongtang** [comment]
Thanks @alextp for the review. After looking through the API,  the PR has been changed to explicitly calling `broadcast_to` before the
```
+        maxval = array_ops.broadcast_to(maxval, shape)
+        minval = array_ops.broadcast_to(minval, shape)
         result = math_ops.add(result * (maxval - minval), minval, name=name)
```

Since `broadcast_to` will return invalid argument error if maxval or minval is larger than the original shape, this will fix the issue #34363, and avoid the restriction of minval and maxval (could be any shape as long as broadcast makes sense and will not 'overflow' after broadcast).

The PR has been updated. Please take a look and see if this works.

**Review by alextp — DISMISSED** [review]

**Comment by gbaned** [comment]
Here are the internal errors, @yongtang can you please verify ?

Traceback (most recent call last):
  File "/third_party/py/absl/third_party/unittest3_backport/case.py", line 37, in testPartExecutor
    yield
  File "/third_party/py/absl/third_party/unittest3_backport/case.py", line 162, in run
    testMethod()
  File "/learning/brain/contrib/slim/model_analyzer_test.py", line 72, in testNoBias
    model_analyzer.analyze_ops(g))
  File "<embedded stdlib>/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "<embedded stdlib>/unittest/case.py", line 506, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 3202 != 4000L

**Comment by yongtang** [comment]
@gbaned @alextp sorry for the late reply. 

The internal test failure is not very straightforward. I am wondering if is it possible to share the code of the internal test  `testNoBias` in file `/learning/brain/contrib/slim/model_analyzer_test.py`?

I assume this is a test related to `tf.contrib.slim`. Given `tf.contrib.slim` has been removed in 2.x, I searched the history of 1.x branch in tensorflow. However, I am not able to locate the `model_analyzer_test.py` file. There is only a `model_analyzer.py` file which might be related. I also could not find a `testNoBias`.

If there are some code available I think I might be quickly find out the reason for the failure.

My wild guess is that the failure is caused by the fact that the following actually adds two more nodes in the graph:
```
        maxval = array_ops.broadcast_to(maxval, shape)
        minval = array_ops.broadcast_to(minval, shape)
```

But without code it is really hard to say what will be the right fix.

**Comment by alextp** [comment]
It should be safe to disable this test internally.

**Comment by yongtang** [comment]
@gbaned Any update on this PR? From the comment https://github.com/tensorflow/tensorflow/pull/38585#issuecomment-624133713 I think the internal test failure could be just ignored?

/cc @alextp

**Review by rthadur — APPROVED** [review]

File: tensorflow/python/kernel_tests/random/random_ops_test.py [hunk 2]
```diff
@@ -413,6 +414,14 @@ def testSingleSessionGraphSeedNotConstant(self):
...
             graph_seed=965)
 
+  def testUniformWithInvalidMaxMindShape(self):
+    # Test case for GitHub issue 34363.
+    with self.assertRaisesRegexp(
+        errors.InvalidArgumentError,
+        "must be no greater than rank of output shape"):
+      array = array_ops.zeros(shape=(1,))
+      random_ops.random_uniform(shape=(), minval=array)
+
```
Comment (rthadur) [inline tensorflow/python/kernel_tests/random/random_ops_test.py:424]: `tensorflow/python/framework/ops.py", line 1820, in _create_c_op
    c_op = pywrap_tf_session.TF_FinishOperation(op_desc)
google3.third_party.tensorflow.python.framework.errors_impl.InvalidArgumentError: Shape must be at most rank 0 but is rank 1 for '{{node random_uniform/BroadcastTo_1}} = BroadcastTo[T=DT_FLOAT, Tidx=DT_INT32](zeros, random_uniform/shape)' with input shapes: [1], [0].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<embedded stdlib>/unittest/case.py", line 59, in testPartExecutor
    yield
  File "<embedded stdlib>/unittest/case.py", line 605, in run
    testMethod()
  File "tensorflow/python/framework/test_util.py", line 1785, in decorated
    return func(self, *args, **kwargs)
  File "/tensorflow/python/kernel_tests/random/random_ops_test.py", line 425, in testUniformWithInvalidMaxMindShape
    random_ops.random_uniform(shape=(), minval=array)
  File "//tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "tensorflow/python/ops/random_ops.py", line 312, in random_uniform
    minval = array_ops.broadcast_to(minval, shape)
  File "/tensorflow/python/ops/gen_array_ops.py", line 854, in broadcast_to
    "BroadcastTo", input=input, shape=shape, name=name)
  File "/tensorflow/python/framework/op_def_library.py", line 744, in _apply_op_helper
    attrs=attr_protos, op_def=op_def)
  File "//tensorflow/python/framework/ops.py", line 3493, in _create_op_internal
    op_def=op_def)
  File "/tensorflow/python/framework/ops.py", line 1983, in __init__
    control_input_ops, op_def)
  File "//tensorflow/python/framework/ops.py", line 1823, in _create_c_op
    raise ValueError(str(e))
ValueError: Shape must be at most rank 0 but is rank 1 for '{{node random_uniform/BroadcastTo_1}} = BroadcastTo[T=DT_FLOAT, Tidx=DT_INT32](zeros, random_uniform/shape)' with input shapes: [1], [0].`

**Review by rthadur — CHANGES_REQUESTED** [review]
@yongtang can you please check the internal error ?

**Comment by yongtang** [comment]
Thanks @rthadur. I think the issue is caused by different error checking in graph vs eager mode so the exception message is slightly different. I have updated the PR that should cover both (I think). Please give it a try and let me know if the issue still persists.

**Review by rthadur — DISMISSED** [review]

**Comment by rthadur** [comment]
@yongtang this test is still failing internally , please check 
`tensorflow/python/framework/test_util.py", line 1785, in decorated
    return func(self, *args, **kwargs)
  File "//tensorflow/python/kernel_tests/random/random_ops_test.py", line 423, in testUniformWithInvalidMaxMindShape
    random_ops.random_uniform(shape=(), minval=array)
  File "//tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)`

**Comment by yongtang** [comment]
Thanks @rthadur Is it possible that you can post the full output of the error message? I think you might be only posting a part of it and it is hard to see where the error is.

**Comment by rthadur** [comment]
Sure , here it is 
`/tensorflow/python/framework/test_util.py", line 1785, in decorated
    return func(self, *args, **kwargs)
  File "/tensorflow/python/kernel_tests/random/random_ops_test.py", line 423, in testUniformWithInvalidMaxMindShape
    random_ops.random_uniform(shape=(), minval=array)
  File "//tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "//tensorflow/python/ops/random_ops.py", line 312, in random_uniform
    minval = array_ops.broadcast_to(minval, shape)
  File "/tensorflow/python/ops/gen_array_ops.py", line 836, in broadcast_to
    _ops.raise_from_not_ok_status(e, name)
  File "/tensorflow/python/framework/ops.py", line 6831, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
google3.third_party.tensorflow.python.framework.errors_impl.UnknownError: error running delegate kernel tf.BroadcastTo: Rank of input (1) must be no greater than rank of output shape (0). [Op:BroadcastTo]`

**Comment by yongtang** [comment]
Thanks @rthadur. I have updated the PR with additional test failure covered. Please give it a try and see if it works now.

**Review by alextp — APPROVED** [review]

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.

**Comment by yongtang** [comment]
Thanks all for the help! 👍 🎉

**Comment by mihaimaruseac** [comment]
Hmm, internal testing on private code was broken by this so it got reverted. I'll try to merge manually this week/next one.

**Comment by yongtang** [comment]
@mihaimaruseac Thanks. 👍 Let me know if I can be of any help.

**Pull request closed without merging**